waiter/src/waiter/instance_tracker.clj

```diff
@@ -0,0 +1,99 @@
+(ns waiter.instance-tracker
+  (:require [clojure.core.async :as async]
+            [clojure.set :as set]
+            [clojure.tools.logging :as log]
+            [metrics.timers :as timers]
+            [waiter.correlation-id :as cid]
+            [waiter.metrics :as metrics]
+            [waiter.util.utils :as utils]))
+
+; Events are being handled by all routers in a cluster for resiliency
+(defprotocol InstanceEventHandler
+
+  (handle-instances-event! [this instances]
+    "handles the instances event")
+
+  (state [this include-flags]
+    "returns the state of the handler:
+    {:last-error-time}"))
+
+(defrecord DefaultInstanceFailureHandler [clock
+                                          handler-state]
+
+  InstanceEventHandler
+
+  (handle-instances-event! [this instances]
+    (let [{:keys [clock handler-state]} this]
+      (log/info "newly failed instances" {:new-failed-instances instances})
+      (reset! handler-state {:last-error-time (clock)})))
```

Comment:
Nice, this lets us write integration tests.
It would be good to also track a specific instance, maybe cache X most recent failed instance IDs mapped by timestamp of when we first learnt about it.

Reply:
good idea